libmemcached/tls.c

```diff
@@ -0,0 +1,595 @@
+            return "Failed to load client certificate\n";
+        case MEMCACHED_SSL_CTX_PRIVATE_KEY_LOAD_FAILED:
+            return "Failed to load private key\n";
+        case MEMCACHED_SSL_CTX_INVALID_PROTOCOL:
+            return "Invalid TLS protocol passed, please use one of the followings: TLSv1.2, TLSv1.3\n";
+        case MEMCACHED_SSL_CTX_CIPHERS_LOAD_FAILED:
+            return "Failed to configure ciphers\n";
+        case MEMCACHED_SSL_CTX_PRIVATE_KEY_MISMATCH:
+            return "Private key does not match the public certificate\n";
+        case MEMCACHED_SSL_CTX_PREFER_SERVER_CIPHER_LOAD_FAILED:
+            return "Failed to configure the server's cipher preferences\n";
+        default:
+            return "Unknown error code\n";
+    }
+}
+
+/**
+ * SSL Connection initialization.
+ */
+static memcached_return_t init_ssl_connection(memcached_instance_st *server, memcached_SSL_CTX *memc_ssl_ctx) {
+    int rv = 0;
+    SSL *ssl = NULL;
+    memcached_SSL *memc_ssl = NULL;
+    SSL_CTX *ssl_ctx = NULL;
+    const char* err_msg = NULL;
+
+    if (!memc_ssl_ctx || !memc_ssl_ctx->ctx) {
+        return memcached_set_error(*server, MEMCACHED_TLS_ERROR, MEMCACHED_AT, memcached_literal_param("Couldn't initialize SSL, memcached_SSL_CTX or SSL_CTX is null"));
+    }
+
+    if (server->privctx) {
+        return memcached_set_error(*server, MEMCACHED_TLS_ERROR, MEMCACHED_AT, memcached_literal_param("The server was already associated with SSL context"));
+    }
+
+    memc_ssl = (memcached_SSL*)libmemcached_calloc(server->root, 1, sizeof(memcached_SSL));
+    if (memc_ssl == NULL) {
+        return memcached_set_error(*server, MEMCACHED_MEMORY_ALLOCATION_FAILURE, MEMCACHED_AT, memcached_literal_param("Out of memory"));
+    }
+
+    // store the current IO functions
+    memc_ssl->default_io_funcs = (void *)server->io_funcs;
+    // set TLS IO functions
+    server->io_funcs = &context_ssl_funcs;
+
+    ssl_ctx = memc_ssl_ctx->ctx;
+    ssl = SSL_new(ssl_ctx);
+    if (!ssl) {
+        rv = ERR_peek_error();
+        goto error;
+    }
+
+    SSL_set_mode(ssl, SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER);
+    WATCHPOINT_ASSERT(server->fd != INVALID_SOCKET);
+    if (!SSL_set_fd(ssl, server->fd)) {
+        err_msg = "Failed to set FD";
+        goto error;
+    }
+    SSL_set_connect_state(ssl);
+
+    if (memc_ssl_ctx->hostname) {
+//#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+        if (!SSL_set1_host(ssl, memc_ssl_ctx->hostname)) {
+            rv = ERR_peek_error();
+            goto error;
+        }
+//#else
+//        X509_VERIFY_PARAM *param = SSL_get0_param(ssl);
+//        /* Enable automatic hostname checks */
+//        X509_VERIFY_PARAM_set_hostflags(param, X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
+//        if (!X509_VERIFY_PARAM_set1_host(param, ssl_ctx->hostname, sizeof(ssl_ctx->hostname) - 1)) {
+//            rv = ERR_peek_error();
+//            goto error;
+//        }
+//#endif
```

Comment:
Remove if you don't need this code.

Reply:
It can be later used if we'll want to expend the use to openssl < 1.1.1
do you prefer I remove it?

Reply:
Yes please remove this for now. If we later need to support OpenSSL version below 1.1 you can add it. I don't see the need to do so.

Reply:
removed